src/Rules/Classes/ImpossibleInstanceOfRule.php

```diff
@@ -59,6 +60,14 @@ public function processNode(Node $node, Scope $scope): array
 			return [];
 		}
 
+		if ($scope->isInTrait()
+			&& $node->expr instanceof Node\Expr\Variable
+			&& is_string($node->expr->name)
+			&& $node->expr->name === 'this'
```

Comment:
I tried different ways on how to detect this properly. since `$this` in traits is class-context dependent the inference turns the type into a `NeverType`. 

checking for the `NeverType` here would be too agressive. therefore I figured using a explicit `$this` check would be the most readable way to handle the case in question